polynote-frontend/polynote/cell.js

```diff
@@ -407,6 +423,52 @@ export class CodeCell extends Cell {
     get content() {
         return this.editor.getValue();
     }
+
+    applyEdits(edits) {
+        // can't listen to these edits or they'll be sent to the server again
+        // TODO: is there a better way to silently apply these edits? This seems like a hack; only works because of
+        //       single-threaded JS, which I don't know whether workers impact that assumption (JS)
+        this.applyingServerEdits = true;
+
+        try {
+            const monacoEdits = [];
+            const model = this.editor.getModel();
+            edits.forEach((edit) => {
+                if (edit.isEmpty()) {
+                    return;
+                }
+
+                const pos = model.getPositionAt(edit.pos);
+                if (edit instanceof messages.Delete) {
+                    const endPos = model.getPositionAt(edit.pos + edit.length);
+                    monacoEdits.push({
+                        range: new monaco.Range(pos.lineNumber, pos.column, endPos.lineNumber, endPos.column),
+                        text: null
+                    });
+                } else if (edit instanceof messages.Insert) {
+                    monacoEdits.push({
+                        range: new monaco.Range(pos.lineNumber, pos.column, pos.lineNumber, pos.column),
+                        text: edit.content,
+                        forceMoveMarkers: true
+                    });
+                }
+            });
+
+            //this.editor.getModel().applyEdits(monacoEdits);
+            // TODO: above API call won't put the edits on the undo stack. Should other people's edits go on my undo stack?
+            //       below calls implement that. It is weird to have other peoples' edits in your undo stack, but it
+            //       also gets weird when they aren't – your undos get out of sync with the document.
+            //       Maybe there's something different that can be done with undo stops – I don't really know what they
+            //       are because it's not well documented (JS)
```

Comment:
Hmm, what does google docs do? We can probably use gdocs as our standard for this as that's what people will be used to.

Reply:
I'm not sure, actually. I'll experiment with it.

For now I'm allowing the foreign edits to go on the undo stack, because the behavior is *much* more broken if they don't, and there doesn't seem to be an easy path to working around that with monaco. We'll see if people complain about it :grin:

Reply:
👍 that's what user testing is for 😉